job-ingest: add ingest module, flux job command, libjob library

garlick opened the pull request:
This PR resurrects #1543.

The interface in job.h (under libjob) for submitting work is:
```c
typedef uint64_t flux_jobid_t;

/* Submit a job to the system.
 * J should be RFC 14 jobspec signed by flux_sign_wrap(), provided
 * flux was built --with-flux-security.  If not, then J should be bare jobspec.
 * Currently the 'flags' parameter must be set to 0.
 * The system assigns a jobid and returns it in the response.
 */
flux_future_t *flux_job_add (flux_t *h, const char *J, int flags);

/* Parse jobid from response to flux_job_add() request.
 * Returns 0 on success, -1 on failure with errno set - and an extended
 * error message may be available with flux_future_error_string().
 */
int flux_job_add_get_id (flux_future_t *f, flux_jobid_t *id);
```

The `flux job` command currently has two subcommands:
```
Usage: flux-job id [OPTIONS] [id ...]
Convert jobid(s) to another form
  -f, --from=dec|kvs-active|words
                         Convert jobid from specified form
  -h, --help             Display this message.
  -t, --to=dec|kvs-active|words
                         Convert jobid to specified form
```
and
```
Usage: flux-job submitbench [OPTIONS] jobspec
Run job(s)
  -R, --reuse-signature  Sign jobspec once and reuse the result for multiple
                         RPCs
  -c, --security-config=pattern
                         Use non-default security config glob
  -f, --fanout=N         Run at most N RPCs in parallel
  -h, --help             Display this message.
  -r, --repeat=N         Run N instances of jobspec
  -s, --sign-type=TYPE   Use non-default mechanism type to sign J
```

There is an opt-in configure option for building with flux-security.  If building without, jobs can be submitted and they are tagged with the submitting user, but they are not signed and thus will only be usable if we implement a short-circuit that allows the instance owner to launch its own work without signature verification.
```
  --with-flux-security    Build with flux-security
```

grondo commented:
Just curious why not `flux_job_submit()` instead of `flux_job_add()`. Are you reserving `flux_submit_submit()` function for a future, higher-level interface? Apologies if I missed something  obvious, I've only just started reviewing.

garlick commented:
Honestly I can't remember.  Happy to change it if that makes more sense.

grondo commented:
> Honestly I can't remember. Happy to change it if that makes more sense.

Well, it is only my opinion but a `flux_job_submit()` call would probably be what users of the API would be expecting...

garlick commented:
>  `flux_job_submit() call would probably be what users of the API would be expecting...`

Yes will change.  Not sure what I was on about there.

garlick commented:
OK, just renamed "add" to "submit" both in the API, the RPC request topic, and the event topic.  I went ahead and squashed that down.

grondo commented:
The "program" to test for `HAVE_FLUX_SECURITY` won't work for the FLUX_TEST_INSTALLED case in general -- though it does work for the case we're concerned about, which is to install the built flux-core sources and then test against the resulting installation directory.

It does feel like maybe we'd want a way outside of compiling a C program to determine if an installed flux-core was built with flux-security support though... maybe we should annotate `flux version` output with compile time information like some programs do? e.g.
```
$ flux version --verbose
flux-core-0.9.0
+security
-caliper
```

or similar?

grondo commented:
Also, if built without flux-security, should the `flux job submitbench` `--reuse-signature` and `--security-config` options be hidden?

garlick commented:
Thanks!  I opened #1632 to track the `flux version --verbose` suggestion and I'll address the other two shortly.

grondo commented:
Now that options are suppressed in `flux job submitbench` when built without flux-security, the built program that returns 1 or 0 if flux-core is built with security or not could be replaced by a grep of the `flux job submitbench --help` output.

garlick commented:
Good point!  OK, I dropped the little proggie and updated the sharness test to look at the output of `flux job --help` instead.

I also took the opportunity to rebase on current master.

If this is looking close, LMK and I'll squash the incremental development.

garlick commented:
I'm going to go ahead and squash.

garlick commented:
oops, just remembered there are two sharness scripts that need updating to check `flux-job --help`.  I'll do that and then squash.

garlick commented:
OK, I think this can go in - could somebody press the button?

SteVwonder commented:
Thanks @garlick!

garlick commented:
Thanks!
